bus-mapping/src/circuit_input_builder.rs

```diff
@@ -1,14 +1,84 @@
 //! This module contains the CircuitInputBuilder, which is an object that takes
 //! types from geth / web3 and outputs the circuit inputs.
-use crate::eth_types::{self, Address, GethExecStep, GethExecTrace};
+use crate::eth_types::{self, Address, GethExecStep, GethExecTrace, Word};
 use crate::evm::GlobalCounter;
 use crate::evm::OpcodeId;
 use crate::exec_trace::OperationRef;
+use crate::geth_errors::*;
 use crate::operation::container::OperationContainer;
 use crate::operation::{Op, Operation};
 use crate::{BlockConstants, Error};
 use core::fmt::Debug;
 
+/// Out of Gas errors by opcode
+#[derive(Debug, PartialEq)]
+pub enum OogError {
+    /// Out of Gas for opcodes which have non-zero constant gas cost
+    Constant,
+    /// Out of Gas for opcodes MLOAD, MSTORE, MSTORE8, CREATE, RETURN, REVERT,
+    /// which have pure memory expansion gas cost
+    PureMemory,
+    /// Out of Gas for SHA3
+    Sha3,
+    /// Out of Gas for CALLDATACOPY
+    CallDataCopy,
+    /// Out of Gas for CODECOPY
+    CodeCopy,
+    /// Out of Gas for EXTCODECOPY
+    ExtCodeCopy,
+    /// Out of Gas for RETURNDATACOPY
+    ReturnDataCopy,
+    /// Out of Gas for LOG
+    Log,
+    /// Out of Gas for CALL
+    Call,
+    /// Out of Gas for CALLCODE
+    CallCode,
+    /// Out of Gas for DELEGATECALL
+    DelegateCall,
+    /// Out of Gas for CREATE2
+    Create2,
+    /// Out of Gas for STATICCALL
+    StaticCall,
+}
+
+/// EVM Execution Error
+#[derive(Debug, PartialEq)]
+pub enum ExecError {
+    /// Always returned for REVERT
+    ExecutionReverted,
+    /// Invalid Opcode
+    InvalidOpcode,
+    /// For opcodes who push more than pop
+    StackOverflow,
+    /// For opcodes which pop, DUP and SWAP, which peek deeper element directly
+    StackUnderflow,
+    /// Out of Gas
+    Oog(OogError),
+    /// For SSTORE, LOG0, LOG1, LOG2, LOG3, LOG4, CREATE, CALL, CREATE2,
+    /// SELFDESTRUCT
+    WriteProtection,
+    /// For CALL, CALLCODE, DELEGATECALL, STATICCALL
+    Depth,
+    /// For CALL, CALLCODE
+    InsufficientBalance,
+    /// For CREATE, CREATE2
+    ContractAddressCollision,
+    /// contract must not begin with 0xef due to EIP #3541 EVM Object Format
+    /// (EOF)
+    InvalidCode,
+    /// For JUMP, JUMPI
+    InvalidJump,
+    /// For RETURNDATACOPY
+    ReturnDataOutOfBounds,
+    /// Internal calculation of gas overflow
+    GasUintOverflow,
```

Comment:
I think in circuit we would verify (treat) this in `Oog` case, otherwise every possible `OogError` could happen as `GasUintOverflow`. Do you think it make sense?

Reply:
In geth (and I'm not sure if in the EVM spec as well), the gas variable is an uint64, and hence can suffer overflows that must be detected when calculating gas costs of dynamic-gas opcodes.
For sure a GasUintOverflow implies an Out of Gas; so if on the circuit it's simpler to only handle Out of Gas (and internally calculate a gas cost that may be higher than u64, or handle a gas cost overflow as Out of Gas), then we can report Oog.

Reply:
Right, I didn't consider that parser needs to also provide the `gas_cost` for prover to argue it's indeed out of gas. Maybe when we are adding gas related field in `ExecStep`, we can have the field `gas_left` as `u64` but `gas_cost` as `F` or `U256` to support such case?

Reply:
I'm wondering: could there be a situation where the gas cost overflows an uint256?

Reply:
hmm, like memory expansion to some 131-bits size, the `gas_cost` following the cost formula could indeed exceed 256-bits (it follows a quadratic formula).

Let me consider more on circuit side and get back to this topic, we probably need different strategies to verify OOG when the value is in different range.

Reply:
But wouldn't this be over the maximum gas_limit of the block already?? How can it overflow a u256? And if it can, doed it really matter?

Reply:
> But wouldn't this be over the maximum gas_limit of the block already?? How can it overflow a u256? And if it can, doed it really matter?

It would be over the maximum gas_limit of the block indeed.  The overflow would happen when calculating the gas_cost, and the circuit needs to prove that the gas_cost was calculated correctly, and from the result prove that it was either small enough to avoid an out of gas, or too high and produces an out of gas; but we should be careful with overflow when calculating this value!  If the circuit internally represents the gas_cost as a field element and it overflows it, then the circuit can't prove that the gas_cost was too high, because it would have wrapped around!

Reply:
Ohh makes complete sense!! Thanks for the explanation!!!!!